core/src/main/java/com/netflix/conductor/core/metadata/MetadataMapperService.java

```diff
@@ -0,0 +1,166 @@
+ * - Immutable definitions within a workflow execution with the added benefit of guaranteeing consistency at runtime.
+ * - Stress is reduced on the storage layer
+ */
+@Singleton
+public class MetadataMapperService {
+
+    public static final Logger logger = LoggerFactory.getLogger(MetadataMapperService.class);
+
+    private final MetadataDAO metadataDAO;
+
+    @Inject
+    public MetadataMapperService(MetadataDAO metadataDAO) {
+        this.metadataDAO = metadataDAO;
+    }
+
+    public WorkflowDef lookupForWorkflowDefinition(String name, Integer version) {
+        Optional<WorkflowDef> potentialDef =
+                version == null ? lookupLatestWorkflowDefinition(name) : lookupWorkflowDefinition(name, version);
+
+        //Check if the workflow definition is valid
+        WorkflowDef workflowDefinition = potentialDef
+                .orElseThrow(() -> {
+                            logger.error("There is no workflow defined with name {} and version {}", name, version);
+                            return new ApplicationException(
+                                    ApplicationException.Code.NOT_FOUND,
+                                    String.format("No such workflow defined. name=%s, version=%s", name, version)
+                            );
+                        }
+                );
+        return workflowDefinition;
+    }
+
+    @VisibleForTesting
+    Optional<WorkflowDef> lookupWorkflowDefinition(String workflowName, int workflowVersion) {
+        // TODO: Update to use ServiceUtils once this is merged with dev
+        // FIXME: Add messages.
+        checkNotNull(workflowName);
+        checkArgument(StringUtils.isNotBlank(workflowName));
+        checkArgument(workflowVersion > 0);
+
+        return metadataDAO.get(workflowName, workflowVersion);
+    }
+
+    @VisibleForTesting
+    Optional<WorkflowDef> lookupLatestWorkflowDefinition(String workflowName) {
+        checkNotNull(workflowName, "Workflow name cannot be null when searching for a definition");
+        checkArgument(StringUtils.isNotBlank(workflowName), "Workflow name cannot be blank when searching for a definition");
+
+        return metadataDAO.getLatest(workflowName);
+    }
+
+    public Workflow populateWorkflowWithDefinitions(Workflow workflow) {
+
+        WorkflowDef workflowDefinition = Optional.ofNullable(workflow.getWorkflowDefinition())
+                .orElseGet(() -> {
+                    WorkflowDef wd = lookupForWorkflowDefinition(workflow.getWorkflowName(), workflow.getWorkflowVersion());
+                    workflow.setWorkflowDefinition(wd);
+                    return wd;
+                });
+
+        workflowDefinition.collectTasks().stream().forEach(
+                workflowTask -> {
+                    if (shouldPopulateDefinition(workflowTask)) {
+                        workflowTask.setTaskDefinition(metadataDAO.getTaskDef(workflowTask.getName()));
```

Comment:
Only call metadataDAO if taskDefinition is not set?

Reply:
Never mind, it is already done in shouldPopulateDefinition method.

Reply:
Only if not set and taskType is SIMPLE